### PR TITLE
GCC: add Apple M1 support

### PR DESCRIPTION
**Pull request by adamjstewart**
Successfully builds on macOS 12.3.1 and Apple M1 Pro with Apple Clang 13.1.6 (builds in under 30 min too!)

See https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/gcc.rb for inspiration

### CODE REVIEW

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 1]
```diff
@@ -254,8 +254,8 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
     conflicts('+bootstrap', when='@:11.1 %apple-clang@12.0.5')
 
     # aarch64/M1 is supported in GCC 12+
-    conflicts('@:11', when='target=aarch64: platform=darwin',
-              msg='Only GCC 12 and newer support macOS M1 (aarch64)')
+    conflicts('@:11.2,11.3.1:', when='target=aarch64: platform=darwin',
+              msg='Only GCC 11.3.0 supports macOS M1 (aarch64)')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/gcc/package.py:260]: GCC 12 was supposed to have M1 support but that feature was delayed. As far as I know, the master branch still doesn't have M1 support, so until that happens we should restrict it to the patched 11.3.0 version.

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 2]
```diff
@@ -290,6 +290,12 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
             patch('https://github.com/iains/gcc-darwin-arm64/commit/20f61faaed3b335d792e38892d826054d2ac9f15.patch?full_index=1',
                   sha256='c0605179a856ca046d093c13cea4d2e024809ec2ad4bf3708543fc3d2e60504b', when='@11.2.0')
 
+        # Apple M1 support, created from branch of Darwin maintainer for GCC:
+        # https://github.com/iains/gcc-11-branch
+        patch('https://raw.githubusercontent.com/Homebrew/formula-patches/22dec3fc/gcc/gcc-11.3.0-arm.diff',
+              sha256='e02006b7ec917cc1390645d95735a6a866caed0dfe506d5bef742f7862cab218',
+              when='@11.3.0 target=aarch64:')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/gcc/package.py:299]: Should we also add `platform=darwin`?

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/gcc/package.py (hunk 2 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/gcc/package.py:299]: This whole set of patches is wrapped in a `if self.platform == 'darwin':` statement. We could instead add that to the `when='` argument if you want.

**Comment by certik** [comment]
I can confirm that this PR builds gcc for me on M1 under 3min 30s:
```console
$ spack install gcc@11.3.0~bootstrap build_type=Release
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/gnuconfig-2021-08-14-xdpppw6jlmcbrrh2f5c6xg66g2ihqx3r
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/zlib-1.2.12-kv442a3m27eyisk3msyutprlor6xfqrc
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/zstd-1.5.2-dxbaafbkwk4qy5yhq2tu3rwzz7xlxsso
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/pkgconf-1.8.0-uh5h4bycgmywlybcmr62y4t4y7ondy73
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/berkeley-db-18.1.40-5usq264s76xutp576hm2uqt4dkzc5cgq
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/autoconf-archive-2022.02.11-fqg62p3v6nmu7hohrgwkjtw6evgqhhlp
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/libiconv-1.16-26iu5wtp44m62c2jnzna5eatwhwbvtsv
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/libsigsegv-2.13-xo4nv7x2kbkd47u3dz7hignhg2iyuiot
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/ncurses-6.2-aht4ig4mbxauojpxs4uogjlap7h34tri
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/diffutils-3.8-tuwwxzqfpqqv7cfjqyza2yrkvayjaxlz
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/m4-1.4.19-l66jpltvffdt252445ackundgtl6fxco
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/readline-8.1-53jxduotneikikbxywzd5yeymx2lmrmk
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/bzip2-1.0.8-w3zybjlargtnz7b3sw3d7vnt7mdh5pxj
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/libtool-2.4.7-gtr6g2nnra7uwsbxnv6tn53lyb7fcsuw
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/gdbm-1.19-shds7kgp7a76h3gzmewh2u46vodbhnrh
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/perl-5.34.1-6xb44vafmqlhtvhbgrangohkf7cokdzp
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/texinfo-6.5-l4sxa45cfrugxk5hmu5onzmgm5id5unc
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/autoconf-2.69-hd5r75lyj5rpeb4l6wyzsookp4t7mxb6
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/automake-1.16.5-iqxpfpx7rht7sbtvdromejth3begn2nl
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/gmp-6.2.1-aodpqorlgcqg7d4lx3s233xeoidepoze
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/mpfr-4.1.0-jk2wmlzcxooad5ynizeqyw4hb5ov2tgg
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/gawk-5.1.1-qxhi2fddvekz4rmerrtvmka2rt2r7imd
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/mpc-1.2.1-466uqr2feqgpqgyzaewxzkdbcsbfojjp
==> Installing gcc-11.3.0-wiug7nsdvmtmlzc5whzmrv6enfmxhjt4
==> No binary for gcc-11.3.0-wiug7nsdvmtmlzc5whzmrv6enfmxhjt4 found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/b4/b47cf2818691f5b1e21df2bb38c795fac2cfbd640ede2d0a5e1c89e338a3ac39.tar.xz
==> Fetching https://raw.githubusercontent.com/Homebrew/formula-patches/22dec3fc/gcc/gcc-11.3.0-arm.diff
==> Applied patch https://raw.githubusercontent.com/Homebrew/formula-patches/22dec3fc/gcc/gcc-11.3.0-arm.diff
==> Ran patch() for gcc
==> gcc: Executing phase: 'autoreconf'
==> gcc: Executing phase: 'configure'
==> gcc: Executing phase: 'build'
==> gcc: Executing phase: 'install'
==> gcc: Successfully installed gcc-11.3.0-wiug7nsdvmtmlzc5whzmrv6enfmxhjt4
  Fetch: 2m 0.23s.  Build: 3m 27.38s.  Total: 5m 27.61s.
[+] /Users/ondrej/repos/spack/opt/spack/darwin-monterey-m1/apple-clang-13.0.0/gcc-11.3.0-wiug7nsdvmtmlzc5whzmrv6enfmxhjt4
```

**Review by michaelkuhn — APPROVED** [review]
LGTM and builds for multiple people

**Comment by sethrj** [comment]
Needs clearance from @alalazo

**Review by alalazo — APPROVED** [review]
Let's add the patch like this, we can remove the conditionals on `sys.platform` later.

**Pull request merged**